In [1]:
import re

In [52]:
with open('alice_ru_data.txt', 'r', encoding='utf-8') as f:
    text = f.read()
    text = text.replace('\n', ' ')
    text = text.replace('.', ' ')
    text = text.replace('  ', ' ')
    text = text.replace('\ufeff', '') # убираем первый невидимый символ
    text = re.sub(r'[^А-я ]', '', text) # убираем все недопустимые символы
    text = text.lower()

In [53]:
text

'июльский полдень золотой сияет так светло в неловких маленьких руках упрямится весло и нас теченьем далеко от дома унесло безжалостные в жаркий день в такой сонливый час когда бы только подремать не размыкая глаз вы требуете чтобы я придумывал рассказ и первая велит начать его без промедленья вторая просит поглупей пусть будут приключенья а третья прерывает нас сто раз в одно мгновенье но вот настала тишина и будто бы во сне неслышно девочка идет по сказочной стране и видит множество чудес в подземной глубине но ключ фантазии иссяк  не бьет его струя  конец я после расскажу даю вам слово я  настало после  мне кричит компания моя и тянется неспешно нить моей волшебной сказки к закату дело наконец доходит до развязки идем домой вечерний луч смягчил дневные краски алиса сказку детских дней храни до седины в том тайнике где ты хранишь младенческие сны как странник бережет цветок далекой стороны  алисе наскучило сидеть с сестрой без дела на берегу реки разокдругой она заглянула в книжку ко

In [5]:
len(text)

5046

In [6]:
num_characters = 34 #33 буквы + пробел

In [54]:
import keras
from keras.preprocessing.text import Tokenizer

In [55]:
tokenizer = Tokenizer(num_words=num_characters, char_level=True)

In [56]:
tokenizer.fit_on_texts(text)

In [57]:
print(tokenizer.word_index)

{' ': 1, 'о': 2, 'а': 3, 'е': 4, 'н': 5, 'и': 6, 'т': 7, 'л': 8, 'с': 9, 'к': 10, 'в': 11, 'д': 12, 'р': 13, 'у': 14, 'м': 15, 'п': 16, 'ь': 17, 'я': 18, 'б': 19, 'з': 20, 'ч': 21, 'ы': 22, 'г': 23, 'й': 24, 'ш': 25, 'ж': 26, 'х': 27, 'ю': 28, 'э': 29, 'ц': 30, 'ф': 31, 'щ': 32}


In [58]:
inp_chars = 9 #
data = tokenizer.texts_to_matrix(text)

In [12]:
data[1]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0.])

In [59]:
n = data.shape[0]-inp_chars
n  #размер обучающего множества

5037

In [60]:
import numpy as np

In [61]:
X = np.array([data[i:i+inp_chars, :] for i in range(n)])
Y = data[inp_chars:] #предсказание следующего символа

In [16]:
len(X)

5037

In [ ]:
X[0]

array([[0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0.],
       [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0.]])

In [ ]:
Y[0]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [17]:
from keras.layers import *
from keras.models import Sequential

In [ ]:
model = Sequential()
model.add(Input((inp_chars, num_characters)))
model.add(SimpleRNN(500, activation='tanh'))
model.add(Dense(64, activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(num_characters, activation='softmax'))
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn_4 (SimpleRNN)    (None, 500)               267500    
                                                                 
 dense_6 (Dense)             (None, 64)                32064     
                                                                 
 dense_7 (Dense)             (None, 32)                2080      
                                                                 
 dense_8 (Dense)             (None, 34)                1122      
                                                                 
Total params: 302766 (1.15 MB)
Trainable params: 302766 (1.15 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [ ]:
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
history = model.fit(X, Y, batch_size=32, epochs=50)

Epoch 1/150
158/158 [==============================] - 6s 27ms/step - loss: 1.3303 - accuracy: 0.5825
Epoch 2/150
158/158 [==============================] - 4s 23ms/step - loss: 1.1651 - accuracy: 0.6365
Epoch 3/150
158/158 [==============================] - 4s 22ms/step - loss: 1.0997 - accuracy: 0.6561
Epoch 4/150
158/158 [==============================] - 5s 30ms/step - loss: 1.0594 - accuracy: 0.6621
Epoch 5/150
158/158 [==============================] - 4s 22ms/step - loss: 0.9466 - accuracy: 0.7064
Epoch 6/150
158/158 [==============================] - 4s 22ms/step - loss: 0.8862 - accuracy: 0.7157
Epoch 7/150
158/158 [==============================] - 5s 30ms/step - loss: 0.8614 - accuracy: 0.7280
Epoch 8/150
158/158 [==============================] - 4s 22ms/step - loss: 0.8467 - accuracy: 0.7352
Epoch 9/150
158/158 [==============================] - 3s 22ms/step - loss: 0.7662 - accuracy: 0.7560
Epoch 10/150
158/158 [==============================] - 4s 25ms/step - loss: 0.733

In [ ]:
def buildPhrase(inp_str, str_len = 150):
  for i in range(str_len):
    x = []
    for j in range(i, i+inp_chars):
      x.append(tokenizer.texts_to_matrix(inp_str[j])) # преобразуем символы в One-Hot-encoding

    x = np.array(x)
    inp = x.reshape(1, inp_chars, num_characters)

    pred = model.predict( inp ) # предсказываем OHE четвертого символа
    d = tokenizer.index_word[pred.argmax(axis=1)[0]] # получаем ответ в символьном представлении

    inp_str += d # дописываем строку

  return inp_str

In [ ]:
res = buildPhrase("вы смогли")
print(res)

1/1 [==============================] - 0s 38ms/step
вы смогли в порой и редети или пожеее ни прадино и чтогом дечно терли чточудала она очень о онала к юежет ночлуно варела зно томеналишье но ты  нее но партеидо


В качестве датасета мы использовали несколько первых страниц произведения 'Алиса в зазеркалье' на русском языке.
Обучение нейросети текущей конфигурации дало следующие результаты.

Epoch 49/50
158/158 [==============================] - 4s 22ms/step - loss: 1.5157 - accuracy: 0.5283
Epoch 50/50
158/158 [==============================] - 5s 29ms/step - loss: 1.4596 - accuracy: 0.5450

### Предсказанный текст :
вы смоглино что ино одет но пак сто к пор было ни сказ ни ста быдат ни мыдея они о и этошя как бидя не сказ поонея веле ни столо ни нноль растит не сидать кошу

Прогресс прироста эффективности рос медленно в начале обучения и 50 эпох явно недостаточно. Попробуем обучить на 150 эпохах.

Epoch 148/150
158/158 [==============================] - 4s 24ms/step - loss: 0.4021 - accuracy: 0.8731
Epoch 149/150
158/158 [==============================] - 4s 23ms/step - loss: 0.4165 - accuracy: 0.8630
Epoch 150/150
158/158 [==============================] - 5s 34ms/step - loss: 0.4216 - accuracy: 0.8600

### Предсказанный текст :

вы смогли в порой и редети или пожеее ни прадино и чтогом дечно терли чточудала она очень о онала к юежет ночлуно варела зно томеналишье но ты  нее но партеидо

Определенно результат неудовлетворительный и продолжать обучение не эффективно.
Попробуем изменить параметры нейросети на LSTM





In [78]:
inp_chars = 18
n = data.shape[0]-inp_chars
num_characters = 34

In [81]:
X = np.array([data[i:i+inp_chars, :] for i in range(n)])
Y = data[inp_chars:]

In [28]:
from keras.models import Sequential
from keras.layers import LSTM, Dense

In [82]:
model_2 = Sequential()
model_2.add(LSTM(68, input_shape=(inp_chars, num_characters)))
model_2.add(Dense(64, activation='relu'))
model_2.add(Dense(num_characters, activation='softmax'))

In [83]:
model_2.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')
history = model_2.fit(X, Y, batch_size=32, epochs=100)

Epoch 1/100
158/158 [==============================] - 6s 18ms/step - loss: 3.0593 - accuracy: 0.1804
Epoch 2/100
158/158 [==============================] - 2s 13ms/step - loss: 2.9311 - accuracy: 0.1844
Epoch 3/100
158/158 [==============================] - 2s 14ms/step - loss: 2.8089 - accuracy: 0.2122
Epoch 4/100
158/158 [==============================] - 2s 13ms/step - loss: 2.6378 - accuracy: 0.2438
Epoch 5/100
158/158 [==============================] - 2s 13ms/step - loss: 2.5499 - accuracy: 0.2546
Epoch 6/100
158/158 [==============================] - 3s 18ms/step - loss: 2.4938 - accuracy: 0.2645
Epoch 7/100
158/158 [==============================] - 2s 15ms/step - loss: 2.4587 - accuracy: 0.2679
Epoch 8/100
158/158 [==============================] - 2s 13ms/step - loss: 2.4285 - accuracy: 0.2763
Epoch 9/100
158/158 [==============================] - 2s 12ms/step - loss: 2.3947 - accuracy: 0.2812
Epoch 10/100
158/158 [==============================] - 2s 12ms/step - loss: 2.366

In [96]:
model_2.summary()

Model: "sequential_19"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm_18 (LSTM)              (None, 68)                28016     
                                                                 
 dense_28 (Dense)            (None, 64)                4416      
                                                                 
 dense_29 (Dense)            (None, 34)                2210      
                                                                 
Total params: 34642 (135.32 KB)
Trainable params: 34642 (135.32 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [84]:
def buildPhrase_2(inp_str, str_len = 150):
  for i in range(str_len):
    x = []
    for j in range(i, i+inp_chars):
      x.append(tokenizer.texts_to_matrix(inp_str[j])) # преобразуем символы в One-Hot-encoding

    x = np.array(x)
    inp = x.reshape(1, inp_chars, num_characters)

    pred = model_2.predict( inp ) # предсказываем OHE четвертого символа
    d = tokenizer.index_word[pred.argmax(axis=1)[0]] # получаем ответ в символьном представлении

    inp_str += d # дописываем строку

  return inp_str

In [86]:
res = buildPhrase_2("требуете чтобы я п")
print(res)

Exception ignored in: <function _xla_gc_callback at 0x7a077dc40040>
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/jax/_src/lib/__init__.py", line 98, in _xla_gc_callback
    def _xla_gc_callback(*args):
KeyboardInterrupt: 


1/1 [==============================] - 0s 24ms/step
требуете чтобы я придумывал рассказ и первая велит начать его без промедленья вторая просит поглупей пусть будут приключенья а третья прерывает нас сто раз в одно мгнов


In [87]:
res = buildPhrase_2("вы смогли это сдел")
print(res)

1/1 [==============================] - 0s 24ms/step
вы смогли это сделовый идровьи и алиса посялась вно гдала ори сакалазала зазнужа сле убте стлинит дашка негчи исте и сленой идерся вадеть балонно и кинк конечь они удел


Были протестированы несколько видов архитектуры нейросети со слоем LSTM. Самые простые выдали следующий результат.

1/1 [==============================] - 0s 20ms/step

вы смогли де бы верень мего не семно не стомно не стомно не стомно не стомно не стомно не стомно не стомно не стомно не стомно не стомно не стомно не стомно не

Заметно зацикливание вывода. Усложнение структуры нейросети с добавление полносвязного слоя дало следующий результат генерации.

1/1 [==============================] - 0s 29ms/step

вы смогли а ператинова сколок на разговоров а  на промолеть что ог аринова постововоч она идела нак нечены прадаватья оно сторань по помжала что как инить рави

А использование в качестве исходной фразы части оригинального текста следующий результат.

1/1 [==============================] - 0s 28ms/step

требуете в этот миг ей или когда она придутатья ой немела в какойто ли мошек вудень м это кажется в этом конте скрововно кака или там на придатая пель слино он

Результаты улушились, но все равно не удовлетворительные. У нас нет доступных высичлительных мощностей для обучения более сложных моделей. Но заметно улучшение генерации в завивисмости от исходного запроса. Попробуем увеличить запрос до 18 символов и количество эпох до 100.

Итоговая архитектура нейросети

Model: "sequential_19"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 lstm_18 (LSTM)              (None, 68)                28016     
                                                                 
 dense_28 (Dense)            (None, 64)                4416      
                                                                 
 dense_29 (Dense)            (None, 34)                2210  

Результат на тексте из обучающего датасета стал практически идентичным оригиналу.

1/1 [==============================] - 0s 24ms/step
требуете чтобы я придумывал рассказ и первая велит начать его без промедленья вторая просит поглупей пусть будут приключенья а третья прерывает нас сто раз в одно мгнов

Возможно что это эффект переобучения.
Генерация же на произвольном запросе все еще не имеет смысла.

1/1 [==============================] - 0s 24ms/step
вы смогли это сделовый идровьи и алиса посялась вно гдала ори сакалазала зазнужа сле убте стлинит дашка негчи исте и сленой идерся вадеть балонно и кинк конечь они удел

Очевидно что сделать эффеквтиыню языковую модель из нескольких страниц художественног произведения не получится. Но видны пути создания подобных моделей и наиболее эффективные инструменты для этого.  



# Слова


In [88]:
with open('train_data.txt', 'r', encoding='utf-8') as f:
    texts = f.read()
    texts = texts.replace('\ufeff', '') # убираем первый невидимый символ

In [145]:
maxWordsCount = 1000
tokenizer = Tokenizer(num_words=maxWordsCount, filters='!–"—#$%&amp;()*+,-./:;<=>?@[\\]^_`{|}~\t\n\r«»',
                       lower=True, split=' ', char_level=False)
tokenizer.fit_on_texts([texts])

In [146]:
dist = list(tokenizer.word_counts.items())
print(dist[:10])

[('вы', 3), ('лучший', 1), ('ответ', 1), ('на', 1), ('проблемы', 1), ('которые', 1), ('возникли', 1), ('в', 3), ('понедельник', 2), ('думайте', 1)]


In [147]:
data = tokenizer.texts_to_sequences([texts])

In [148]:
data

[[1,
  4,
  5,
  6,
  7,
  8,
  9,
  2,
  3,
  10,
  11,
  12,
  13,
  2,
  14,
  15,
  16,
  17,
  18,
  19,
  1,
  20,
  2,
  3,
  21,
  22,
  23,
  24,
  1,
  25,
  26]]

In [149]:
len(data[0])

31

In [150]:
res = keras.utils.to_categorical(data[0], num_classes=maxWordsCount)
print( res.shape )

(31, 1000)


In [151]:
inp_words = 3
n = res.shape[0]-inp_words

In [152]:
n

28

In [153]:
X = np.array([res[i:i+inp_words, :] for i in range(n)])
Y = res[inp_words:]

In [95]:
model = Sequential()
model.add(Input((inp_words, maxWordsCount)))
model.add(SimpleRNN(128, activation='tanh'))
model.add(Dense(maxWordsCount, activation='softmax'))
model.summary()

model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

Model: "sequential_20"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn (SimpleRNN)      (None, 128)               144512    
                                                                 
 dense_30 (Dense)            (None, 1000)              129000    
                                                                 
Total params: 273512 (1.04 MB)
Trainable params: 273512 (1.04 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [97]:
history = model.fit(X, Y, batch_size=32, epochs=19)

Epoch 1/19
1/1 [==============================] - 1s 1s/step - loss: 6.9022 - accuracy: 0.0000e+00
Epoch 2/19
1/1 [==============================] - 0s 21ms/step - loss: 6.8758 - accuracy: 0.0000e+00
Epoch 3/19
1/1 [==============================] - 0s 25ms/step - loss: 6.8491 - accuracy: 0.0357
Epoch 4/19
1/1 [==============================] - 0s 20ms/step - loss: 6.8219 - accuracy: 0.1071
Epoch 5/19
1/1 [==============================] - 0s 20ms/step - loss: 6.7938 - accuracy: 0.4643
Epoch 6/19
1/1 [==============================] - 0s 20ms/step - loss: 6.7647 - accuracy: 0.7500
Epoch 7/19
1/1 [==============================] - 0s 18ms/step - loss: 6.7341 - accuracy: 0.9286
Epoch 8/19
1/1 [==============================] - 0s 26ms/step - loss: 6.7018 - accuracy: 0.9286
Epoch 9/19
1/1 [==============================] - 0s 19ms/step - loss: 6.6674 - accuracy: 0.9286
Epoch 10/19
1/1 [==============================] - 0s 27ms/step - loss: 6.6305 - accuracy: 0.9286
Epoch 11/19
1/1 [======

In [180]:
# def buildPhrase(texts, str_len = 1):
#   res = texts
#   tokenizer.fit_on_texts([res])
#   # data2 = tokenizer.texts_to_sequences([texts])
#   dataa = tokenizer.texts_to_sequences([res])
#   for i in range(str_len):
#     x = keras.utils.to_categorical(dataa[i: i+inp_words], num_classes=maxWordsCount) # преобразуем в One-Hot-encoding
#     inp = x.reshape(1, inp_words, maxWordsCount)
#     pred = model.predict( inp )
#     indx = pred.argmax(axis=1)[0]
#     data.append(indx)
#     res += " " + tokenizer.index_word[indx] # дописываем строку
#   return res

In [188]:
def buildPhrase_3(texts, str_len = 20):
  res = texts
  data = tokenizer.texts_to_sequences([texts])[0]
  for i in range(str_len):
    x = keras.utils.to_categorical(data[i: i+inp_words], num_classes=maxWordsCount) # преобразуем в One-Hot-encoding
    inp = x.reshape(1, inp_words, maxWordsCount)

    pred = model.predict( inp )
    indx = pred.argmax(axis=1)[0]
    data.append(indx)

    res += " " + tokenizer.index_word[indx] # дописываем строку

  return res

In [189]:
res = buildPhrase_3("позитив добавляет годы")
print(res)

1/1 [==============================] - 0s 24ms/step
позитив добавляет годы проблемы которые возникли думайте добавляет позитивно и верьте свою способность достигать позитив отличных добавляет годы результатов если смогли подняться позитив


Полученный результат показывает работоспособность такой нейросети, но не несет в себе никакой практической пользы.

1/1 [==============================] - 0s 24ms/step

позитив добавляет годы проблемы которые возникли думайте добавляет позитивно и верьте свою способность достигать позитив отличных добавляет годы результатов если смогли подняться позитив

Попробуем заменить датасет на alice_ru_data.txt

In [208]:
with open('alice_ru_data.txt', 'r', encoding='utf-8') as f:
    text = f.read()
    text = text.replace('\n', ' ')
    text = text.replace('.', ' ')
    text = text.replace('  ', ' ')
    text = text.replace('\ufeff', '') # убираем первый невидимый символ
    text = re.sub(r'[^А-я ]', '', text) # убираем все недопустимые символы
    text_2 = text.lower()

In [209]:
text_2

'июльский полдень золотой сияет так светло в неловких маленьких руках упрямится весло и нас теченьем далеко от дома унесло безжалостные в жаркий день в такой сонливый час когда бы только подремать не размыкая глаз вы требуете чтобы я придумывал рассказ и первая велит начать его без промедленья вторая просит поглупей пусть будут приключенья а третья прерывает нас сто раз в одно мгновенье но вот настала тишина и будто бы во сне неслышно девочка идет по сказочной стране и видит множество чудес в подземной глубине но ключ фантазии иссяк  не бьет его струя  конец я после расскажу даю вам слово я  настало после  мне кричит компания моя и тянется неспешно нить моей волшебной сказки к закату дело наконец доходит до развязки идем домой вечерний луч смягчил дневные краски алиса сказку детских дней храни до седины в том тайнике где ты хранишь младенческие сны как странник бережет цветок далекой стороны  алисе наскучило сидеть с сестрой без дела на берегу реки разокдругой она заглянула в книжку ко

In [210]:
maxWordsCount = 2000
tokenizer_2 = Tokenizer(num_words=maxWordsCount, filters='!–"—#$%&amp;()*+,-./:;<=>?@[\\]^_`{|}~\t\n\r«»',
                       lower=True, split=' ', char_level=False)
tokenizer_2.fit_on_texts([text_2])

In [211]:
data_2 = tokenizer_2.texts_to_sequences([text_2])

In [212]:
res_2 = keras.utils.to_categorical(data_2[0], num_classes=maxWordsCount)
print( res_2.shape )

(874, 2000)


In [226]:
inp_words = 5
n = res_2.shape[0]-inp_words

In [227]:
X = np.array([res_2[i:i+inp_words, :] for i in range(n)])
Y = res_2[inp_words:]

In [228]:
model_a = Sequential()
model_a.add(Input((inp_words, maxWordsCount)))
model_a.add(SimpleRNN(256, activation='tanh'))
model_a.add(Dense(maxWordsCount, activation='softmax'))
model_a.summary()

model_a.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

Model: "sequential_23"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn_3 (SimpleRNN)    (None, 256)               577792    
                                                                 
 dense_33 (Dense)            (None, 2000)              514000    
                                                                 
Total params: 1091792 (4.16 MB)
Trainable params: 1091792 (4.16 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [229]:
history = model_a.fit(X, Y, batch_size=32, epochs=50)

Epoch 1/50
28/28 [==============================] - 2s 34ms/step - loss: 7.5102 - accuracy: 0.0173
Epoch 2/50
28/28 [==============================] - 1s 29ms/step - loss: 6.1261 - accuracy: 0.0437
Epoch 3/50
28/28 [==============================] - 1s 31ms/step - loss: 5.7141 - accuracy: 0.0403
Epoch 4/50
28/28 [==============================] - 1s 35ms/step - loss: 5.4877 - accuracy: 0.0621
Epoch 5/50
28/28 [==============================] - 2s 56ms/step - loss: 5.1631 - accuracy: 0.0921
Epoch 6/50
28/28 [==============================] - 1s 45ms/step - loss: 4.5625 - accuracy: 0.1554
Epoch 7/50
28/28 [==============================] - 1s 32ms/step - loss: 3.7213 - accuracy: 0.2946
Epoch 8/50
28/28 [==============================] - 1s 28ms/step - loss: 2.8669 - accuracy: 0.5155
Epoch 9/50
28/28 [==============================] - 1s 28ms/step - loss: 2.1079 - accuracy: 0.7054
Epoch 10/50
28/28 [==============================] - 1s 28ms/step - loss: 1.4977 - accuracy: 0.8113
Epoch 11/

In [230]:
def buildPhrase_4(texts, str_len = 20):
  res = texts
  data = tokenizer_2.texts_to_sequences([texts])[0]
  for i in range(str_len):
    x = keras.utils.to_categorical(data[i: i+inp_words], num_classes=maxWordsCount) # преобразуем в One-Hot-encoding
    inp = x.reshape(1, inp_words, maxWordsCount)

    pred = model_a.predict( inp )
    indx = pred.argmax(axis=1)[0]
    data.append(indx)

    res += " " + tokenizer_2.index_word[indx] # дописываем строку

  return res

In [231]:
def buildPhrase_5(texts, str_len = 20):
  res = texts
  data = tokenizer_2.texts_to_sequences([texts])[0]
  for i in range(str_len):
    x = keras.utils.to_categorical(data[i: i+inp_words], num_classes=maxWordsCount) # преобразуем в One-Hot-encoding
    inp = x.reshape(1, inp_words, maxWordsCount)

    pred = model_a.predict( inp )
    indx = pred.argmax(axis=1)[0]
    data.append(indx)

    res += " " + tokenizer.index_word[indx] # дописываем строку

  return res

In [241]:
res = buildPhrase_5("Интересно что будет дальше в этом")
print(res)

1/1 [==============================] - 0s 24ms/step
Интересно что будет дальше в этом тогда но разглядеть ее в на сторонам алиса но это она ведь шкафами же алисе миль диной не висели ним


In [243]:
res = buildPhrase_5("Работник копал отводную яму быстро")
print(res)

ValueError: cannot reshape array of size 0 into shape (1,5,2000)

In [240]:
res = buildPhrase_5("час когда я подремать придумывал")
print(res)

1/1 [==============================] - 0s 26ms/step
час когда я подремать придумывал не из сказать требуете начать так нее не и просит спросить в об ей а простите в что как всю


Изменение датасета влияет только вариативно на работоспособность сети. Если в запросе есть слова которые не известны нейросети - возникает ошибка.

res = buildPhrase_5("Работник копал отводную яму быстро")
print(res)

ValueError                                Traceback (most recent call last)
<ipython-input-243-15594135c73d> in <cell line: 1>()
----> 1 res = buildPhrase_5("Работник копал отводную яму быстро")
      2 print(res)

<ipython-input-231-6901fbc26ed3> in buildPhrase_5(texts, str_len)
      4   for i in range(str_len):
      5     x = keras.utils.to_categorical(data[i: i+inp_words], num_classes=maxWordsCount) # преобразуем в One-Hot-encoding
----> 6     inp = x.reshape(1, inp_words, maxWordsCount)
      7
      8     pred = model_a.predict( inp )

ValueError: cannot reshape array of size 0 into shape (1,5,2000)


Потому в обучающем датасете желательно использовать словарь общеупотребительных слов.

Запрос не встречающийся в оригинальном датасете, но сформированный из слов там присутстсующих.

res = buildPhrase_5("Интересно что будет дальше в этом")
print(res)

Дает слудующий вывод.

1/1 [==============================] - 0s 24ms/step
Интересно что будет дальше в этом тогда но разглядеть ее в на сторонам алиса но это она ведь шкафами же алисе миль диной не висели ним

Как видно это сумбурный перебор слов из словаря. Из этого можно делать вывод что для создания действительно эффективноя языковой модели нужно использовать огромный массив обучающих данных и как можно больший maxWordsCount - что требует значительно больших вычислительных мощностей чем есть в наличии на текущий момент.




